src/evaluation_metrics.py

```diff
@@ -0,0 +1,89 @@
+import numpy as np
+
+
+def mapk(k, image):
+    """
+    Computes the mean average precision of an image for a k set of predictions
+    Args:
+        - k: number of predictions
+        - image: image used to compute the k most similar images
+    Returns:
+        - mapk: mean average precision of the image for a k set of predictions
+    """
+    actual = [image]
+    predicted = [k_retrieval(image)] # llista de llistes dels k predits
+    return mmetrics.mapk(k, actual, predicted)
+
+def global_mapk(k, images):
+    """
+    Computes the mean of the mean average precision of a list of images for a k set of predictions of each image
+    Args:
+        - k: number of predictions
+        - images: list of images used to compute the k most similar images
+    Returns:
+        - mean_mapk: mean of the mean average precision of the images for a k predictions
+    """
+    mean_mapk = 0.0
+    for i in images:
+        mean_mapk += mapk(k, i)
+    return mean_mapk / float(len(images))
+
+
+def binarized_image_comparision(image_ref, image_pred):
+    """
+    Computes the precision, recall and f1 score between a binarized image and its prediciton
+    Args:
+        - image_ref: gound truth image as numpy.array
+        - image_pred: predicted image as numpy.array
+    Returns:
+        - precision: precision of the predicted image in comparision to the original image
+        - recall: recall of the predicted image in comparision to the original image
+        - f1: f1 score of the predicted image in comparision to the original image
+    """
+    tp = 0.0
+    tn = 0.0
+    fp = 0.0
+    fn = 0.0
+    for i in range(image_ref.shape[0]):
+        for j in range(image_ref.shape[1]):
+            if(image_ref[i][j] and image_pred[i][j]):
+                tp+=1.0
+            elif(image_ref[i][j] and not(image_pred[i][j])):
+                fn+=1.0
+            elif(not(image_ref[i][j]) and image_pred[i][j]):
+                fp+=1.0
+            else:
+                tn+=1.0
+    precision = tp / (tp + fp)
+    recall = tp / (tp + fn)
+    f1 = 2 * precision * recall / (precision + recall)
+    return precision, recall, f1
+
+def global_binarized_image_comparisions(refs_images, preds_images):
+    """
+    Computes the precision, recall and f1 score between a list of binarized image and its predicitons
+    Args:
+        - refs_images: list of gound truth images
+        - preds_images: list of predicted images
+    Returns:
+        - precision: mean of the precision of the predicted images in comparision to the original images
+        - recall: mean of the recall of the predicted images in comparision to the original images
+        - f1: mean of the f1 score of the predicted images in comparision to the original images
+    """
+    precision = 0.0
+    recall = 0.0
+    f1 = 0.0
+    for i in range(len(refs_images)):
+        precision_aux, recall_aux, f1_aux = compute_image_comparision(refs_images[i], preds_images[i])
```

Comment:
The function is defined in line 32.

In order to be efficient it has to be done with 4 functions or 1 function (because then you dont need to recompute the tp, tn, fp and fn rates every time). And i decided to do it with one because i think that we will always need all this infomation together. But if you prefer i can slpit it in 4 functions insted of 1. Well, 4 functions for the image by image comparision + 3 functions for the global comparisions, so 7 functions.

Reply:
You're right :)